File: backends/node-nestjs/src/audit.ts
import { randomUUID } from "node:crypto";
import type { Queryable } from "./db.js";

/**
 * The append-only audit trail (SPEC rule 18): every moderator/admin mutation
 * writes an entry; there is no code path that updates or deletes one. Callers
 * inside a transaction pass their client so the entry commits atomically with
 * the change it records.
 */
export async function recordAudit(
  db: Queryable,
  actor: string,
  action: string,
  targetType: string,
  targetId: string,
  detail?: Record<string, unknown>,
): Promise<void> {
  await db.query(
    `insert into audit_entries (id, actor, action, target_type, target_id, detail, created_at)
     values ($1, $2, $3, $4, $5, $6::jsonb, $7)`,
    [randomUUID(), actor, action, targetType, targetId, detail ? JSON.stringify(detail) : null, new Date()],
  );
}

File: backends/python-django/stackverse_api/audit.py
from __future__ import annotations

from typing import Any

from .models import AuditEntry
from .time import now_utc


def record_audit(
    actor: str, action: str, target_type: str, target_id: str, detail: dict[str, Any] | None = None
) -> None:
    AuditEntry.objects.create(
        actor=actor,
        action=action,
        target_type=target_type,
        target_id=str(target_id),
        detail=detail,
        created_at=now_utc(),
    )
